### PR TITLE
#20887: Reduce GC pressure by compacting to reduce long lived garbage

### PR DESCRIPTION
**Pull request by drewhk**
Reduce potentially long lived garbage by compacting ByteStrings in DelimiterFraming and Rechunker

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala [hunk 1]
```diff
@@ -193,7 +193,7 @@ object Framing {
         } else if (buffer.slice(possibleMatchPos, possibleMatchPos + separatorBytes.size) == separatorBytes) {
           // Found a match
           val parsedFrame = buffer.slice(0, possibleMatchPos).compact
-          buffer = buffer.drop(possibleMatchPos + separatorBytes.size)
+          buffer = buffer.drop(possibleMatchPos + separatorBytes.size).compact
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala]: we can also do this after a certain count of emitted frames if necessary

**Comment by drewhk** [comment]
Fixes #20887

**Comment by patriknw** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala (hunk 1 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala]: Good idea, though let's keep it simple for now. Unless someone finds this dominating their usecase let's keep it as is. Properly doing that would require a FramingAttributes etc, a bit much work for here/right now

**Comment by ktoso** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala]: I don't think it is needed. I used to use the heuristic that I record the position of the last compaction, and then compact once the position of the first byte of the buffer is larger than the last position (i.e. there is no overlap anymore with the original ByteString after the last compaction).

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala]: Cool, I see

**Comment by johanandren** [comment]
LGTM

**Pull request merged**